kinode/src/vfs.rs

```diff
@@ -594,26 +593,49 @@ async fn open_file<P: AsRef<Path>>(
     truncate: bool,
 ) -> Result<Arc<Mutex<fs::File>>, VfsError> {
     let path = path.as_ref().to_path_buf();
-    Ok(match open_files.get(&path) {
-        Some(file) => file.value().0.clone(),
-        None => {
+
+    if open_files.len() >= MAX_OPEN_FILES {
+        cleanup_files(&open_files);
+    }
+
+    // avoid race conditions
+    match open_files.entry(path.clone()) {
+        Entry::Occupied(mut entry) => {
+            entry.get_mut().1 = Instant::now(); // update last access time
+            Ok(entry.get().0.clone())
+        }
+        Entry::Vacant(entry) => {
             let file = Arc::new(Mutex::new(
                 tokio::fs::OpenOptions::new()
                     .read(true)
                     .write(true)
                     .create(create)
                     .truncate(truncate)
                     .open(&path)
-                    .await
-                    .map_err(|e| VfsError::IOError {
-                        error: e.to_string(),
-                        path: path.display().to_string(),
-                    })?,
+                    .await?,
             ));
-            open_files.insert(path, (file.clone(), Instant::now()));
-            file
+            entry.insert((file.clone(), Instant::now()));
+            Ok(file)
         }
-    })
+    }
+}
+
+fn cleanup_files(open_files: &DashMap<PathBuf, (Arc<Mutex<fs::File>>, Instant)>) {
+    let current_count = open_files.len();
+    let target_count = max(current_count / 2, MAX_OPEN_FILES / 2);
+
+    // collect all entries, sort by last access time
+    let mut entries: Vec<_> = open_files
+        .iter()
+        .map(|entry| (entry.key().clone(), entry.value().1))
+        .collect();
+    entries.sort_by(|a, b| a.1.cmp(&b.1));
+
+    // remove oldest entries until we reach the target count
+    for (path, _) in entries.into_iter().take(current_count - target_count) {
+        // use remove_if_present to avoid potential race conditions
+        open_files.remove_if(&path, |_, _| true);
+    }
```

Comment:
is there a way to make this more efficient using `retain` on the dashmap with a fixed duration calculated by the average of the map? could be better than cloning every path. something like

1. calculate average last access time by traversing map once
2. retain entries only with a below average last access time

Reply:
this is a good point, I'll try something out!